metaflow/cmd/code/__init__.py

```diff
@@ -0,0 +1,226 @@
+    output: bool = bool(kwargs_dict.pop("output", False))
+    op_args: Mapping[str, Any] = {**kwargs_dict}
+    return perform_diff(tmpdir, target_dir=target_dir, output=output, **op_args)
+
+
+def op_pull(tmpdir: str, dst: str, **kwargs: Mapping[str, Any]) -> None:
+    if os.path.exists(dst):
+        echo(f"❌  Directory *{dst}* already exists")
+    else:
+        shutil.move(tmpdir, dst)
+        echo(f"Code downloaded to *{dst}*")
+
+
+def op_patch(tmpdir: str, dst: str, **kwargs: Mapping[str, Any]) -> None:
+    diffs = perform_diff(tmpdir, output=True) or []
+    with open(dst, "w", encoding="utf-8") as f:
+        for out in diffs:
+            out = out.replace(tmpdir, "/.")
+            out = out.replace("+++ b/./", "+++ b/")
+            out = out.replace("--- b/./", "--- b/")
+            out = out.replace("--- a/./", "--- a/")
+            out = out.replace("+++ a/./", "+++ a/")
+            f.write(out)
+    echo(f"Patch saved in *{dst}*")
+    path = run(
+        ["git", "rev-parse", "--show-prefix"], text=True, stdout=PIPE
+    ).stdout.strip()
+    if path:
+        diropt = f" --directory={path.rstrip('/')}"
+    else:
+        diropt = ""
+    echo("Apply the patch by running:")
+    echo_always(
+        f"git apply --verbose{diropt} {dst}", highlight=True, bold=True, err=True
+    )
+
+
+@code.command()
+@click.argument("metaflow-run")
+def diff(metaflow_run: str, **kwargs: Mapping[str, Any]) -> None:
+    """
+    Do a 'git diff' of the current directory and a Metaflow run.
+    """
+    _ = run_op(metaflow_run, op_diff, **kwargs)
+
+
+@code.command()
+@click.argument("source_run")
+@click.argument("target_run")
+def diff_runs(source_run: str, target_run: str, **kwargs: Mapping[str, Any]) -> None:
+    """
+    Do a 'git diff' between two Metaflow runs.
+    """
+    _ = run_op_diff_runs(source_run, target_run, **kwargs)
+
+
+@code.command()
+@click.argument("metaflow_run")
+@click.option(
+    "--dir", help="Destination directory (default: {runspec}_code)", default=None
+)
+def pull(
+    metaflow_run: str, dir: Optional[str] = None, **kwargs: Mapping[str, Any]
+) -> None:
+    """
+    Pull the code of a Metaflow run.
+    """
+    if dir is None:
+        dir = metaflow_run.lower().replace("/", "_") + "_code"
+    op_args: Mapping[str, Any] = {**kwargs, "dst": dir}
+    run_op(metaflow_run, op_pull, **op_args)
+
+
+@code.command()
+@click.argument("metaflow_run")
+@click.option("--file", help="Patch file name (default: {runspec}.patch", default=None)
+def patch(
+    metaflow_run: str, file: Optional[str] = None, **kwargs: Mapping[str, Any]
+) -> None:
+    """
+    Create a patch file for the current dir with a Metaflow run.
+    """
+    if file is None:
+        file = metaflow_run.lower().replace("/", "_") + ".patch"
+    if confirm_overwrite(file):
+        op_args: Mapping[str, Any] = {**kwargs, "dst": file}
+        run_op(metaflow_run, op_patch, **op_args)
+
+
+def confirm_overwrite(file):
```

Comment:
can this be handled directly through click instead so that `metaflow code patch ... -y` works